### PR TITLE
[nrf fromtree] net: lib: http: Correct http_client edge-case behavior

### PR DESCRIPTION
**Pull request by glarsennordic**
Correct various small edge-case behaviors that have been accidentally introduced in the http_client.

- http_client_req no longer incorrectly returns -ETIMEDOUT on NULL HTTP resonse. -ETIMEDOUT is now only returned when the underlying TLS socket times out.
- http_client_req now returns -ECONRESET upon incomplete (but non-NULL) HTTP response. The request callback is no longer called in this case (as with any other error state).
- http_wait_data has been refactored slightly to increase clarity.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

(cherry picked from commit https://github.com/zephyrproject-rtos/zephyr/pull/70911/commits/30c7d0920c9b03dd15f9ac1913738c290c003275)

### CODE REVIEW

**Comment by glarsennordic** [comment]
@gmarull Is it possible to get this included in the ongoing upmerge? This is a bugfix for IRIS-8638

**Review by plskeggs — APPROVED** [review]

File: subsys/net/lib/http/http_client.c [hunk 1]
```diff
@@ -438,24 +464,28 @@ static int http_wait_data(int sock, struct http_request *req, int32_t timeout)
...
 		} else if (ret < 0) {
+			ret = -errno;
 			goto error;
 		}
 		if (fds[0].revents & (ZSOCK_POLLERR | ZSOCK_POLLNVAL)) {
+			ret = -errno;
 			goto error;
 		} else if (fds[0].revents & ZSOCK_POLLHUP) {
 			/* Connection closed */
 			LOG_DBG("Connection closed");
```
Comment (jayteemo) [inline subsys/net/lib/http/http_client.c:478]: can this and the one below be moved to the `closed:` label?

**Review by jayteemo — APPROVED** [review]

File: subsys/net/lib/http/http_client.c (hunk 1 above)
Comment (glarsennordic) [reply subsys/net/lib/http/http_client.c:478]: Hypothetically, yes, and I started to do that, but I ultimately thought the code was a little easier to read this way.

I don't care a ton either way, but if you also don't care too much I will leave as-is to speed up review/merge

**Comment by rlubos** [comment]
We'd need a manifest PR for this (to verify CI).

**Review by tony-le-24 — APPROVED** [review]

**Comment by glarsennordic** [comment]
Ah! Right. Thank you. I'll make that happen

**Review by jukkar — APPROVED** [review]

**Comment by tony-le-24** [comment]
confirmed that it works on my CI, built and tested locally, the issue was gone.

**Comment by glarsennordic** [comment]
@rlubos looks like CI is green on the manifest PR after latest rebase

**Pull request merged**